text/XXX-annotate-unwind-rust.md

```diff
@@ -0,0 +1,191 @@
+- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: ???
+- Rust Issue: ???
+
+# Summary
+[summary]: #summary
+
+A new function annotation, `#[unwind(Rust)]`, will be introduced; it will
+explicitly permit `extern` functions to unwind (`panic`) without aborting. No
+guarantees are made regarding the specific unwinding implementation.
+
+# Motivation
+[motivation]: #motivation
+
+This will enable resolving
+[rust-lang/rust#58794](https://github.com/rust-lang/rust/issues/58794) without
+breaking existing code.
+
+Currently, unwinding through an FFI boundary is always undefined behavior. We
+would like to make the behavior safe by aborting when the stack is unwound to
+an FFI boundary. However, there are existing Rust crates (notably, wrappers
+around the `libpng` and `libjpeg` C libraries) that make use of the current
+implementation's behavior. The proposed annotation would act as an "opt out" of
+the safety guarantee provided by aborting at an FFI boundary.
+
```

Comment:
I didn't realize you felt this way about the specific impetus for the request and the proposal itself; I had thought, from you last comments in the issue thread, that you were in agreement that this was a reasonable request and a viable proposal.

I disagree with your perspective on the UB issue and am willing to have that conversation, but I'm now wondering if that discussion should be continued back on the `internals` thread so that it doesn't make this PR too noisy. If I had understood that we didn't yet have agreement on the way forward, I wouldn't have opened this PR.

Reply:
(By "the UB issue" I don't mean expecting LLVM not to use `nounwind` for optimization; I agree that doing so is incorrect.)

Reply:
> I didn't realize you felt this way about the specific impetus for the request and the proposal itself; I had thought, from you last comments in the issue thread, that you were in agreement that this was a reasonable request and a viable proposal.

I thought [this comment](https://github.com/rust-lang/rust/issues/58794#issuecomment-486969970) expressed what I believe would be an appropriate motivation for this attribute.

Reply:
@gnzlbg I agree that FFI Rust->Rust is an appropriate motivation for the attribute, but it felt disingenuous to me to advocate for it on that ground alone since it's not the real motivation for myself or @kornelski to be pushing for an annotation of some sort changing FFI unwind behavior. What I took from your phrasing in that comment about Rust->X->Rust was that you agreed it would be appropriate for @kornelski and anyone else to use the proposed annotation for Rust->X->Rust as long as they realize that they're relying on an implementation detail:

> ...by accident, it would allow those doing Rust->X->Rust to continue to "work" as long as their assumptions about implementation details still hold.

Knowledge about implementation details doesn't seem like an overly burdensome requirement for `unsafe` interactions _between languages_; that's how the computing world managed to function without a formally defined C ABI for several decades.

Reply:
> What I took from your phrasing in that comment about Rust->X->Rust was that you agreed it would be appropriate for @kornelski and anyone else to use the proposed annotation for Rust->X->Rust as long as they realize that they're relying on an implementation detail:

I don't see where I agreed to that. My comment says that "panics in Rust->Rust FFI" is a simple extension to the language that solves an useful problem, and that such a feature would allow some of the people that are invoking undefined-behavior in Rust today to continue doing so after the soundness fix for panics in Rust FFI lands. 

I never suggested `#[unwind(Rust)]` as a solution to "panic in Rust->X", nor say that it would be appropriate for anyone to write code with undefined behavior. If anything, I showed multiple times how to write code without UB in stable Rust today, and argued that because that is possible "panics in Rust->X" is not a problem worth solving.